packages/utils/src/joiner/remote-joiner.ts

```diff
@@ -0,0 +1,535 @@
+  MedusaContainer,
+  RemoteJoinerQuery,
+} from "@medusajs/types"
+import GraphQLParser from "./graphq-ast"
+
+interface NestedExpands {
+  [key: string]: {
+    fields: string[]
+    args?: JoinerArgument[]
+    expands?: NestedExpands
+  }
+}
+
+interface ParsedExpand {
+  property: string
+  serviceConfig: JoinerServiceConfig
+  fields: string[]
+  args?: JoinerArgument[]
+  expands?: NestedExpands
+}
+
+const BASE_PATH = "_root"
+export class RemoteJoiner {
+  private serviceConfigs: JoinerServiceConfig[]
+  private serviceConfigCache: Map<string, JoinerServiceConfig> = new Map()
+
+  constructor(
+    private container: MedusaContainer,
+    serviceConfigs: JoinerServiceConfig[]
+  ) {
+    this.serviceConfigs = this.createSelfReferences(serviceConfigs)
+  }
+
+  private createSelfReferences(serviceConfigs: JoinerServiceConfig[]) {
+    for (const service of serviceConfigs) {
+      const propName = service.serviceName.toLowerCase()
+      service.relationships.push({
+        alias: propName,
+        foreignKey: propName + "_id",
+        primaryKey: "id",
+        serviceName: service.serviceName,
+      })
+    }
+
+    return serviceConfigs
+  }
+
+  private findServiceConfig(
+    serviceName: string
+  ): JoinerServiceConfig | undefined {
+    if (!this.serviceConfigCache.has(serviceName)) {
+      const config = this.serviceConfigs.find(
+        (config) => config.serviceName === serviceName
+      )
+      this.serviceConfigCache.set(serviceName, config!)
+    }
+    return this.serviceConfigCache.get(serviceName)
+  }
+
+  private async fetchData(
+    expand: ParsedExpand,
+    pkField: string,
+    ids?: (string | string[])[],
+    relationship?: any
+  ): Promise<any> {
+    let uniqueIds = Array.isArray(ids) ? ids : ids ? [ids] : undefined
+
+    if (uniqueIds) {
+      const isCompositeKey = Array.isArray(uniqueIds[0])
+      if (isCompositeKey) {
+        const seen = new Set()
+        uniqueIds = uniqueIds.filter((idArray) => {
+          const key = JSON.stringify(idArray)
+          const isNew = !seen.has(key)
+          seen.add(key)
+          return isNew
+        })
+      } else {
+        uniqueIds = Array.from(new Set(uniqueIds.flat()))
+      }
+    }
+
+    const serviceConfig = expand.serviceConfig
+    const moduleRegistryName =
+      serviceConfig.serviceName[0].toLowerCase() +
+      serviceConfig.serviceName.slice(1) +
+      "Service"
+    const service = this.container.resolve(moduleRegistryName)
+
+    if (relationship) {
+      pkField = relationship.inverse
+        ? relationship.foreignKey.split(".").pop()!
+        : relationship.primaryKey
+    }
+
+    const methodName = relationship?.inverse
+      ? `getBy${this.toPascalCase(pkField)}`
+      : "list"
+
+    const response = await service[methodName]({
+      fields: expand.fields,
+      args: expand.args,
+      expands: expand.expands,
+      options: {
+        [pkField]: uniqueIds,
+      },
+    })
+
+    const filterFields = (
```

Comment:
todo: Move that to a static method of the class to avoid this method to always re create this util locally